mkosi/distributions/gentoo.py

```diff
@@ -211,125 +118,142 @@ def fetch_fix_stage3(self) -> None:
             else:
                 die("profile names changed upstream?")
 
+        stage3_tar_path = state.cache_dir / stage3_tar
+
         stage3_url_path = urllib.parse.urljoin(
-            self.portage_cfg["GENTOO_MIRRORS"],
-            f"releases/{self.arch}/autobuilds/{stage3_tar}",
+            state.config.mirror, f"releases/{arch}/autobuilds/{stage3_tar}",
         )
-        stage3_tar_path = self.state.cache_dir / stage3_tar
-        stage3_tmp_extract = stage3_tar_path.with_name(stage3_tar.name + ".tmp")
-        if not stage3_tar_path.is_file():
-            log_step(f"Fetching {stage3_url_path}")
-            stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
-            urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
-
-        stage3_tmp_extract.mkdir(parents=True, exist_ok=True)
-
-        with flock(stage3_tmp_extract):
-            if not stage3_tmp_extract.joinpath(".cache_isclean").exists():
-                log_step(f"Extracting {stage3_tar.name} to {stage3_tmp_extract}")
 
+        cls.stage3_cache = state.cache_dir.joinpath("stage3")
+        if not stage3_tar_path.is_file():
+            if cls.stage3_cache.exists():
+                log_step("New stage3 is available, removing the old one")
+                unlink_try_hard(cls.stage3_cache)
+            with complete_step(f"Fetching {stage3_url_path}"):
+                stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
+                urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
+        cls.stage3_cache.mkdir(parents=True, exist_ok=True)
+        if not cls.stage3_cache.joinpath(".cache_isclean").exists():
```

Comment:
Do we really need this file? We can just check if the extracted directory exists no?

Reply:
we make that directory unconditionally since `tar` changes into, `-C`, before extracting.

Reply:
Check if it is non-empty then

Reply:
should be fixed although there might be a more *pythonic* way to do this.